### PR TITLE
SELECT ... FOR UPDATE when looking up groups

### PR DESCRIPTION
**Pull request by usmanm**

### CODE REVIEW

File: src/backend/executor/nodePhysicalGroupLookup.c [hunk 1]
```diff
@@ -45,24 +48,89 @@ ExecInitPhysicalGroupLookup(PhysicalGroupLookup *node, EState *estate, int eflag
...
+	EState *estate = outer->js.ps.state;
+	Buffer buffer;
+	HeapUpdateFailureData hufd;
+	HeapTuple tup;
+	Relation rel;
+	TupleTableSlot *slot = NULL;
+	BitmapHeapScanState *scan;
 
+	/* Get next tuple from subplan, if any. */
+lnext:
```
Comment (loadzero) [inline src/backend/executor/nodePhysicalGroupLookup.c]: Thats an interesting use of goto. What did the non-goto version look like?

**Comment by derekjn** [comment]
Looks good to me :+1:

File: src/backend/executor/nodePhysicalGroupLookup.c (hunk 1 above)
Comment (usmanm) [inline src/backend/executor/nodePhysicalGroupLookup.c]: There wasn't need for any retry before since we didn't do any locking of tuples.

Comment (loadzero) [inline src/backend/executor/nodePhysicalGroupLookup.c]: Sorry, I wasn't being clear. I meant that I find backwards jumping goto harder to read and more error-prone compared to a loop/break/continue construct.

I was wondering whether there was a clarity reason for doing it this way.

Comment (usmanm) [inline src/backend/executor/nodePhysicalGroupLookup.c]: That's how Postgres does it in `nodeLockRows.c`. This logic is pretty similar to what goes inside that.

Comment (loadzero) [inline src/backend/executor/nodePhysicalGroupLookup.c]: Gotcha. Makes sense now.

**Pull request closed without merging**